The job **`x86_64-gnu-pre-stabilization`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/27546779522/job/81422319074) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/81422319074) [(plain)](https://github.com/rust-lang/rust/commit/e7a57aa7df11fcdcc4b7467d8b54ce94bc63c88d/checks/81422319074/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
1 error: size: Size(16 bytes)
2   --> $DIR/layout.rs:11:1
3    |

28 LL | struct ScalableTuple(ScalableU8, ScalableU8, ScalableU8);
29    | ^^^^^^^^^^^^^^^^^^^^
30 
- error: aborting due to 5 previous errors
+ error: aborting due to 10 previous errors
32 
---
To only update this specific test, also pass `--test-args scalable-vectors/layout.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1/bin/rustc" "/checkout/tests/ui/scalable-vectors/layout.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/scalable-vectors/layout" "-A" "unused" "-W" "unused_attributes" "-A" "internal_features" "-A" "incomplete_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers"
stdout: none
--- stderr -------------------------------
error: scalable vectors are not supported on this architecture
##[error]  --> /checkout/tests/ui/scalable-vectors/layout.rs:10:1
   |
---

error: size: Size(16 bytes)
##[error]  --> /checkout/tests/ui/scalable-vectors/layout.rs:11:1
   |
LL | struct ScalableFloat(f32); //~ ERROR: size: Size(16 bytes)
   | ^^^^^^^^^^^^^^^^^^^^

error: size: Size(8 bytes)
##[error]  --> /checkout/tests/ui/scalable-vectors/layout.rs:15:1
   |
LL | struct ScalableU8WithFewerCount(u8); //~ ERROR: size: Size(8 bytes)
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: size: Size(16 bytes)
##[error]  --> /checkout/tests/ui/scalable-vectors/layout.rs:19:1
   |
LL | struct ScalableU8(u8); //~ ERROR: size: Size(16 bytes)
   | ^^^^^^^^^^^^^^^^^

error: size: Size(16 bytes)
##[error]  --> /checkout/tests/ui/scalable-vectors/layout.rs:23:1
   |
LL | struct ScalableBool(bool); //~ ERROR: size: Size(16 bytes)
   | ^^^^^^^^^^^^^^^^^^^

error: size: Size(48 bytes)
##[error]  --> /checkout/tests/ui/scalable-vectors/layout.rs:27:1
   |
LL | struct ScalableTuple(ScalableU8, ScalableU8, ScalableU8); //~ ERROR: size: Size(48 bytes)
   | ^^^^^^^^^^^^^^^^^^^^

error: aborting due to 10 previous errors
------------------------------------------

```

</details>

For more information how to resolve CI failures of this job, visit this [link](https://rustc-dev-guide.rust-lang.org/tests/pre-stabilization-ci-job.html).

--- a/compiler/rustc_abi/src/layout.rs
+++ b/compiler/rustc_abi/src/layout.rs
@@ -1498,19 +1498,21 @@ where
     // Compute the size and alignment of the vector
     let size =
         elt.size.checked_mul(count, dl).ok_or_else(|| LayoutCalculatorError::SizeOverflow)?;
-    let (repr, align) = match kind {
+    let (repr, size, align) = match kind {
         SimdVectorKind::Scalable(number_of_vectors) => (
             BackendRepr::SimdScalableVector { element, count, number_of_vectors },
+            size.checked_mul(number_of_vectors.0 as u64, dl)
+                .ok_or_else(|| LayoutCalculatorError::SizeOverflow)?,
             dl.llvmlike_vector_align(size),
         ),
         // Non-power-of-two vectors have padding up to the next power-of-two.
         // If we're a packed repr, remove the padding while keeping the alignment as close
         // to a vector as possible.
         SimdVectorKind::PackedFixed if !count.is_power_of_two() => {
-            (BackendRepr::Memory { sized: true }, Align::max_aligned_factor(size))
+            (BackendRepr::Memory { sized: true }, size, Align::max_aligned_factor(size))
         }
         SimdVectorKind::PackedFixed | SimdVectorKind::Fixed => {
-            (BackendRepr::SimdVector { element, count }, dl.llvmlike_vector_align(size))
+            (BackendRepr::SimdVector { element, count }, size, dl.llvmlike_vector_align(size))
         }
     };
     let size = size.align_to(align);

--- a/compiler/rustc_target/src/callconv/mod.rs
+++ b/compiler/rustc_target/src/callconv/mod.rs
@@ -755,7 +755,10 @@ impl<'a, Ty> FnAbi<'a, Ty> {
 
             if arg_idx.is_none()
                 && arg.layout.size > Primitive::Pointer(AddressSpace::ZERO).size(cx) * 2
-                && !matches!(arg.layout.backend_repr, BackendRepr::SimdVector { .. })
+                && !matches!(
+                    arg.layout.backend_repr,
+                    BackendRepr::SimdVector { .. } | BackendRepr::SimdScalableVector { .. }
+                )
             {
                 // Return values larger than 2 registers using a return area
                 // pointer. LLVM and Cranelift disagree about how to return

--- a/tests/ui/scalable-vectors/layout.rs
+++ b/tests/ui/scalable-vectors/layout.rs
@@ -1,0 +1,28 @@
+//@ only-aarch64
+#![feature(rustc_attrs)]
+#![crate_type = "lib"]
+
+// Tests that the computed layout size of scalable vectors is equal to
+// `element size * element count * number of vectors`. Scalable vectors are of course scalable and
+// so do not have a fixed size, but using this size with things like `llvm.memcpy` produces the
+// correct and expected results.
+
+#[rustc_dump_layout(size)]
+#[rustc_scalable_vector(4)]
+struct ScalableFloat(f32); //~ ERROR: size: Size(16 bytes)
+
+#[rustc_dump_layout(size)]
+#[rustc_scalable_vector(8)]
+struct ScalableU8WithFewerCount(u8); //~ ERROR: size: Size(8 bytes)
+
+#[rustc_dump_layout(size)]
+#[rustc_scalable_vector(16)]
+struct ScalableU8(u8); //~ ERROR: size: Size(16 bytes)
+
+#[rustc_dump_layout(size)]
+#[rustc_scalable_vector(16)]
+struct ScalableBool(bool); //~ ERROR: size: Size(16 bytes)
+
+#[rustc_dump_layout(size)]
+#[rustc_scalable_vector]
+struct ScalableTuple(ScalableU8, ScalableU8, ScalableU8); //~ ERROR: size: Size(48 bytes)

--- a/tests/ui/scalable-vectors/layout.stderr
+++ b/tests/ui/scalable-vectors/layout.stderr
@@ -1,0 +1,32 @@
+error: size: Size(16 bytes)
+  --> $DIR/layout.rs:12:1
+   |
+LL | struct ScalableFloat(f32);
+   | ^^^^^^^^^^^^^^^^^^^^
+
+error: size: Size(8 bytes)
+  --> $DIR/layout.rs:16:1
+   |
+LL | struct ScalableU8WithFewerCount(u8);
+   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+error: size: Size(16 bytes)
+  --> $DIR/layout.rs:20:1
+   |
+LL | struct ScalableU8(u8);
+   | ^^^^^^^^^^^^^^^^^
+
+error: size: Size(16 bytes)
+  --> $DIR/layout.rs:24:1
+   |
+LL | struct ScalableBool(bool);
+   | ^^^^^^^^^^^^^^^^^^^
+
+error: size: Size(48 bytes)
+  --> $DIR/layout.rs:28:1
+   |
+LL | struct ScalableTuple(ScalableU8, ScalableU8, ScalableU8);
+   | ^^^^^^^^^^^^^^^^^^^^
+
+error: aborting due to 5 previous errors
+

